src/Bicep.Core/Parsing/BaseParser.cs

```diff
@@ -0,0 +1,1185 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Extensions;
+using Bicep.Core.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+
+namespace Bicep.Core.Parsing
+{
+    public abstract class BaseParser
+    {
+        protected readonly TokenReader reader;
+
+        protected readonly ImmutableArray<IDiagnostic> lexerDiagnostics;
+
+        protected BaseParser(string text)
+        {
+            // treating the lexer as an implementation detail of the parser
+            var diagnosticWriter = ToListDiagnosticWriter.Create();
+            var lexer = new Lexer(new SlidingTextWindow(text), diagnosticWriter);
+            lexer.Lex();
+
+            this.lexerDiagnostics = diagnosticWriter.GetDiagnostics().ToImmutableArray();
+
+            this.reader = new TokenReader(lexer.GetTokens());
+        }
+
+        public abstract ProgramSyntax Program();
+
+        protected abstract SyntaxBase Declaration();
+
+        private static bool CheckKeyword(Token? token, string keyword) => token?.Type == TokenType.Identifier && token.Text == keyword;
+
+        private static int GetOperatorPrecedence(TokenType tokenType)
```

Comment:
>GetOperatorPrecedence [](http://example.com/codeflow?start=27&length=21)

All members were added by using the "pull members up to base class" refactor. Didn't make any code changes except for changing `private` to `protected` to get things to compile.